### DISCUSSION TITLE
Making the most of your included free Codespaces usage

### DISCUSSION THREAD

**Discussion by craiglpeters** [community/community#39697] [Codespaces] [answered]
This post has been replaced by the content in https://docs.github.com/en/codespaces/troubleshooting/troubleshooting-included-usage
## Retired post
Since we released Codespaces for All a few weeks ago some of you have had questions about the free usage and how we count usage. See ["About billing for GitHub Codespaces"](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces) for the details, but here's a tl;dr and some tips.

## Delete codespaces when you are done with your task
Because Codespaces are ephemeral and can be quickly spun up for your work, **you can safely delete them once you commit and [push the changes you've made to a branch](https://docs.github.com/en/codespaces/developing-in-codespaces/using-source-control-in-your-codespace#pushing-changes-to-your-remote-repository)**. We recommend that you delete codespaces when you are done with a task to maximize the value of your free included usage, or set a spending limit if you want to keep codespaces around for long periods of time. 

## Codespaces usage
We count two dimensions for Codespaces usage: compute and storage. We first deduct the amount of each you use from your included (free) usage until either compute or storage are consumed. Once one of those limits is reached you can continue to work in your codespace once you've setup a spending limit and a payment method. GitHub provides enough free usage such that most people will never reach the included limits for open source contributions or side projects. It is not enough free usage to do every day work.

**Codespaces compute** is counted in [core hours](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#billing-for-compute-usage), which is the sum of active time in a codespace, times the multiplier for that machine type (e.g. 2 for a 2 core machine, 8 for an 8 core machine, and so on). Active time is the time from create or resume, until suspension. The default timeout (time until suspend) is 30 minutes, but you can change that.

**Codespaces storage** is counted in [GB months](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#billing-for-storage-usage), which is a cumulative measure of the total storage each codespace consumes from creation to deletion plus the storage for prebuilds. See the documentation for a detailed example
- Codespaces consume storage while they exist - [configure your retention period](https://github.com/settings/codespaces) to ensure ones you forget about are deleted automatically - the default value is 30 days. If you were a participant in our beta program for Free and Pro accounts, you may have old codespaces that predate the retention period setting. We have added a retention period to those codespaces which you can see on your [Codespaces index page](https://github.com/codespaces) (link at the top of most GitHub pages)
- Storage is calculated hourly and accumulated to your prior storage total - consumed storage is cumulative for the duration of your month-long billing cycle which means that during the billing period the value you see on your [billing page](https://github.com/settings/billing) increases, and only decreases when a new billing cycle starts and we add more included usage to your account. When you delete a codespace or a prebuild it decreases the rate at which storage usage is accumulated.
- We recently stopped counting our default container against your storage usage. Your storage is not being billed for the size of the default container for any codespaces using the default image created or restarted after Nov 8, 2022.
- Ensure that you are using prebuilds for only as many versions and as many regions as you need (see ["Storage costs for prebuilds"](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#storage-costs-for-prebuilds)).
- Note that if your included storage usage is exhausted, new Codespaces Prebuilds are disabled until you set up a spending limit or your included usage quota is reset.

We are very happy to be bringing Codespaces to all of you, and are grateful for your feedback.

**Comment by aaryno** [comment]
I was using this as an Educator this semester, right before codespaces became "free" - My impression was that I would be able to use this resource as an educator for my approved github organization. Suddenly, with 2 weeks left in the semester, any repo checked out from GitHub Classroom cannot run a codespace. Is this a quirk of how the "free" rollout happened? Will I get my class back on track? 

This is so frustrating. I can't give my personal credit card out to all my students to use for this org and I can't ask them to give up their credit card to finish the last week of this class. I'm just an Adjunct Professor teaching this graduate course in open source computing at a State University. I'm so lost and so frustrated and need help urgently.

**Reply by craiglpeters** [reply]
Apologies for missing this @aaryno. I'll reach out to figure out what is going on

**Comment by tachibana-shin** [comment]
Hi,

I used up my Codespaces free quota on this account last month

But it's now another month and I still haven't received access to Codespaces?

what's up?

**Reply by aprilla** [reply]
Hi @tachibana-shin, you should be able to view https://github.com/settings/billing and under the Codespaces section, find out what quotas you've used and when they will be refreshed.  

Happy to answer any additional questions you might have.

**Reply by tachibana-shin** [reply]
> Hi @tachibana-shin, you should be able to view https://github.com/settings/billing and under the Codespaces section, find out what quotas you've used and when they will be refreshed.
> 
> Happy to answer any additional questions you might have.

Gihub says I've used up 100% of my quota and need to remove the codespaces. But i deleted all of them and github still says i need to delete more. I don't know what to do please help me

**Reply by craiglpeters** [reply]
Codespaces storage is cumulative, so deleting a codespace stops new storage use from being counted for that codespace. Your usage will remain at that level until your quota of included usage is reset. You can always see that at the URL @aprilla provided.
<img width="411" alt="Codespaces quota resets in 3 days" src="https://user-images.githubusercontent.com/9445180/206213876-036375ad-3366-4699-89cc-8b91dd621b3c.png">

**Reply by spenserblack** [reply]
>Storage is calculated hourly and accumulated to your prior storage total

For example, if I create a 1GB codespace, and keep it for 2 hours before deleting it, then that's 2GB of usage for the billing period, right?

**Reply by craiglpeters** [reply]
I've written out [an example in the documentation](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#billing-for-storage-usage). For your example. if your codespace consumes 1 GB of storage, and you run it for 2 hours, and then delete it, it will consume 1 GB/(720 hours in a 30 day month billing period) * (48 hrs in 2 days) ~= 0.07 GB-months of usage.

Note: fixed an error in which I had originally failed to convert 2 days into 48 hours

**Reply by spenserblack** [reply]
Ah, I missed that (very helpful and clear) example, thank you!

**Reply by tachibana-shin** [reply]
> Codespaces storage is cumulative, so deleting a codespace stops new storage use from being counted for that codespace. Your usage will remain at that level until your quota of included usage is reset. You can always see that at the URL @aprilla provided. <img alt="Codespaces quota resets in 3 days" width="411" src="https://user-images.githubusercontent.com/9445180/206213876-036375ad-3366-4699-89cc-8b91dd621b3c.png">

![IMG_20221208_084345](https://user-images.githubusercontent.com/45375496/206335483-3ad84509-0c5e-4bf9-8c54-1430de9ae188.jpg)

I went to that link and github said i used up 15GB. But there is 1 problem that I lost access to codespaces from last month and it seems that my quota for this month is not resetting.

I don't know where the problem is. I was previously given beta codespaces access, is that the reason I can't reset my quota?

Please help me

**Reply by craiglpeters** [reply]
@tachibana-shin your quota for included usage will reset in 4 days (as is indicated on your screenshot) and you'll be able to continue to use Codespaces for free. Your included usage is reset once per month. 

Because you were in the beta, you may have old codespaces that you no longer need that can consume your included storage - you should delete any of those you no longer need from github.com/codespaces.

**Reply by sohaha** [reply]
> @tachibana-shin your quota for included usage will reset in 4 days (as is indicated on your screenshot) and you'll be able to continue to use Codespaces for free. Your included usage is reset once per month.
> 
> Because you were in the beta, you may have old codespaces that you no longer need that can consume your included storage - you should delete any of those you no longer need from github.com/codespaces.

It's strange, mine didn't reset either

**Reply by ankur0101** [reply]
Are free 120 hours counted as one time offer or github resets the hours back to 0 every month?

**Reply by Alice52** [reply]
@craiglpeters 

I donot use codespace for long time, but github said i used up 15GB, what's happened?
![image](https://user-images.githubusercontent.com/42330329/211715679-9580c211-bd04-47e3-8a28-9458039d9b50.png)

**Comment by MichaelRWolf** [comment]
**You've used 100% of included services for GitHub Codespaces Storage**

I signed up for Codespaces a few days before the free plan rolled out.  It's great!

I see _that_ I _accumulated_ enough usage that I'm at 100%.  I can see that on the red bar.

But I don't know _how_ that happened, so I don't know what to do to avoid it.  With only 1 big number (the summary), I can't learn how to manage the details.

Can I get an hour-by-hour detail of how it accumulated?  (Actually, day-by-day would probably be good enough)

Can I get a repo-by-repo detail of how it accumulated?

Better still, since the charges accumulate hourly based on space multiplied by repo, it would be better to get time*repo snapshots (by the hour or by the day).

P.S.  I have a feeling that I should have *deleted* codespaces instead of *stopping* them, but I'd like to *know* (rather than *feel*).  Data will help that.

**Reply by spenserblack** [reply]
(copying from https://github.com/community/community/discussions/38780#discussioncomment-4334857)

I think it would be awesome if, in https://github.com/codespaces, each codespaces had usage stats (`x GiB/hr`) to show how much storage each codespace accumulates hourly. Additionally, since prebuilds take up storage, it would be nice to see that stat somewhere in the repository page, too.

**Reply by craiglpeters** [reply]
Y'all are so right! We are working to get that data to you. We are planning to show you the utilized storage for each codespace on /codespaces soon. Additionally the historical usage reports are in the works for all products, but those will take a bit longer.

**Comment by MichaelRWolf** [comment]
Suggestion:  Warn of excessive usage before it's all gone

Remember when we were charged for text messages?  Remember when it was easy to exceed plan limit for number of voice minutes?

Remember when you'd get a notice saying that you were on target to exceed those?

It sure would be nice to get mid-cycle alerts for Codespace, too.

Something like
> Excessive usage:  You have used %50 of max storage limit in 30% of the billing cycle length.  If you continue at this rate, you will hit the limit 12 days before the end of the billing cycle.

**Comment by MichaelRWolf** [comment]
I do appreciate that you put "We appreciate your feedback".

So... as I incrementally unwedge myself from a deadlock
* Want to know if codespace is up to date w/ origin and on a clean branch
* Can't open it to check.  Must update spending.  Fortunately, I had this set up to support a project, else it would be another road blocking step to add billing info.
* Oh.  It is clean.  **I could have saved myself the bother of opening it.**  I could have just deleted it.

That final line gives me the idea for a user-generated (that's me!) backlog item:
1) If a codespace is safe to delete, say so on the main panel.  (I propose "safe to delete" means that "git status" returns the following, but I may be missing something).  If it's not, say why.  This could be as simple as caching output from "git status" as the codespace is closed (or times out).  I may be wrong or niaive.

<img width="635" alt="image" src="https://user-images.githubusercontent.com/2851070/208178921-b21be17d-4db2-4d79-82fa-513f170844db.png">

2) Provide a "sneak peak" to open codespaces in a restricted mode that would allow inspection to learn if it is OK to delete.  (Yes, this is another HOW to the basic WHAT (see if it's OK to delete closed codespace).  This one is probably harder than previoius suggestion.

**Reply by ChocolateLoverRaj** [reply]
Idk if this was recently added after your suggestion, but I think the dot indicating un-pushed changes solves your problem:
![screenshot](https://user-images.githubusercontent.com/52586855/210880996-34257064-31ce-42c8-b7b0-dc866ad9a395.png)
I like how even though I can't open the codespace, I can still get the changes with the "Export changes to a branch" button:
![image](https://user-images.githubusercontent.com/52586855/210881153-b65a0da6-4612-4697-a3a9-a56a4f1fd161.png)

**Comment by MichaelRWolf** [comment]
Final suggestion (for today...)

Make the following line bold, italics, 4 points bigger, and highlighted in yellow so that the TL;DR; crowd can easily scan for this buried lead
> you can safely delete them once you commit and [push the changes you've made to a branch](https://docs.github.com/en/codespaces/developing-in-codespaces/using-source-control-in-your-codespace#pushing-changes-to-your-remote-repository).

That was my problem!  I left the water running, then went on vacation!

Thanks for a great service!  I'm diggin' it.

**Reply by craiglpeters** [reply]
Thank you @MichaelRWolf! Consolidating my responses, as all of these suggestions quite closely related. For future readers of this discussion, would you mind consolidating them into this one? Or, if we have your permission, a community manager could merge your comments for you.

On your github.com/codespaces page we WILL be providing the storage usage size shortly - I wanted to get that out before the release for everyone, but that just wasn't possible. The historical view of how you used your storage is also in the works, but that report will take a bit longer to get out.

On your /codespaces page you can already see whether your git status is clean. For example from mine you'll note that several of my codespaces have "Uncommitted changes" which means I might want to think twice before deleting them.
<img width="353" alt="git statuses" src="https://user-images.githubusercontent.com/9445180/208474334-bc41e0b6-2b97-45f4-8a02-c8ccb45ba76d.png">
We are working on including more places to suggest deleting, or not deleting, a codespace as a part of workflows. But developers work in so many different ways, we are being careful not to prescribe a BEST workflow.

If you are locked out of your codespaces because you have exhausted all of your included usage, you can always export your changes to a branch so you can continue work elsewhere - that is essentially the "restricted mode" that only allows you to push to a branch.

I like your suggestion on raising the prominance/noticability of the line about deleting codespaces.

Happy Coding!

**Comment by yangsfang** [comment]
Hi Craig, thank you for writing [the example in the documentation](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#billing-for-storage-usage). It has helped immensely to understand Codespaces pricing. But there is still confusion after reading this thread as well as that documentation page. More clarification is desired.

1. We have the ability to choose the machine type as well as changing it after the fact. The publicly available machine types are pre-set with 32GB of storage (see first image below). At any point in time, a Codespace could be consuming less than 32GB. 32GB can be thought of as an upper limit of actual usage (see second image of a `df -h` query). Is the storage charge based on the 32GB limit, or the 13GB disk usage inside the machine, or a measurement against some specific directories?

![image](https://user-images.githubusercontent.com/31492969/208559286-f0d5f094-79c0-4b76-8b25-d24757041c0c.png)
![image](https://user-images.githubusercontent.com/31492969/208559646-2df59cb0-253c-41b9-b95b-7ef060cae2ce.png)

2. If the charge is against the limit of 32GB-month prorated to the hour, then it would not be possible to keep a Codespace inactive for an entire month, because the 15GB-month free quota would run out at about day 14. One must take care to delete the Codespace after each use, which is a hassle. Is there any plan by the Codespaces team to release a machine type of 15GB, such that such machine can be kept inactive for an entire month on free quota, and booted up as needed?
3. If the charge is against the actual usage of 13GB, can that number be surfaced to the Codespaces UI? (If this is what you meant by `On your github.com/codespaces page we WILL be providing the storage usage size shortly` in your previous comment, then ignore this)
4. I poked around with terminal and the 13GB seems to include the VSCode binary. Can these binaries be on a shared image so this does not need to be charged to each Codespace? (I'm thinking in virtual machine terms, there would be a shared static read-only disk with the OS and VSCode binaries, while the cloned workspace and any user installed components would be on a chargeable mutable disk image)
5. Either way, can the Codespaces show the current consumption rate for each Codespace on its main UI? In the example given in the documentation, `one codespace that uses 100 GB of storage and has existed for one hour you will have used 100 / (24 * 30) = 0.1388 GB-months of storage in a 30-day month`, I would love to see the number `0.1388 GB/hour` on the Codespaces UI next to the machine name. This would clearly communicate to the user how much storage is being charged, and they can easily calculate how soon the storage quota would run out and what the paid hourly rate would be, rather than digging into a CSV file inside the billing UI which also does not give an hourly consumption rate.

Thanks.

**Reply by ghost** [reply]
This is the most relevant question for me. I burned through my free monthly usage within 15 days with just a basic dev container with python installed and a tiny repo. I completely understand the free limits are set in a way where trying to use a dev container for daily work will not reach the end of the month and thats not its intended purpose, I'm totally ok with that. But I would like to understand how best to maximum the amount of time I can spend with a dev container, both inside of it, but also just having it available without needing to delete it anytime i wont use it for a few days. Being able to set a storage limit for a dev container would be great for this (assuming that the 32gb size we cannot control would result in a faster monthly usage than being able to set a limit of say 16gb ect)

**Comment by blaticslm** [comment]
"You've used 90% of included services for GitHub Codespaces Storage.“
![图片](https://user-images.githubusercontent.com/47236078/209418524-40539dcb-a7e6-455f-8947-b9144b16fb1c.png)


I am confused there. I don't use any codespace but it says I have used that much. May you explain what is happening behind this? Thank you!

**Reply by bartr** [reply]
I ran into the same issue because I had several pre-builds stored (in multiple regions). That will eat through the free storage in a few days.

A great feature would be to see all of the pre-builds that are taking up storage. Make deleting / editing much easier than having to go to every repo, settings, Codespaces.

I do love that Codespaces is available to everyone now as it's game-changing IMO.

**Reply by blaticslm** [reply]
I agree with you...I have checked all my personal repos, and none of them even goes to 500MB. I have no idea on what way this 15GB can be used up.

**Reply by bartr** [reply]
if you're using pre-builds, the images can be +1 GB. The old default was 4 regions and 3 versions (I think). That's 12 GB per repo per month. I think I had 4 repos setup with pre-build.

**Comment by Choongkyu** [comment]
I removed all of my codespaces at least one cycle ago, but I'm still seeing an issue with not being able to create a new codespace due to my having used up all of my free available storage. I was participating in the beta prior to that so I'm wondering if the issues are tied to my usage not resetting properly.

**Comment by ENate** [comment]
What if I do not have a credit card and still want to pay somehow?

**Comment by Xag8031** [comment]
Trying to use github codespaces for classes during school. I am using github student. I managed to use 160.92 of 180.00 included core hours and I was wondering if I could reduce how much I was using?
Would also like if I could customize the cores and notifications more to my liking

**Reply by ChocolateLoverRaj** [reply]
If u aren't already u can use 2 core machines instead of 4 core machines.

**Reply by joshspicer** [reply]
Yes, please check which "Machine Type" you're creating your codespaces against.  Billing is charged per "core hours", so a 4-core machine will be twice as expensive as a 2-core machine.  Hope that helps - please see https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces for more info on billing.

**Comment by hubwriter** [comment, marked as answer]
We've published a new article on the docs site which should help with questions around the free usage included with personal accounts: https://docs.github.com/en/codespaces/troubleshooting/troubleshooting-included-usage

**Comment by el-buku** [comment]
Anybody knows why 8core machines are now unavailable? I'm a free plan but used to be able to spin up 8 core codespaces on any repo, not anymore. Are they available to pro plan users, or they need to be enabled, similar to the 32gb ones?

**Reply by mnoszczak** [reply]
@el-buku I can access 8 core machines on one account, but not the other (both are upgraded to PRO, so that's probably not the case). 

I think that the documentation here is not really helpful:
> Typically, you can run your codespace on a choice of remote machine types. These machine types offer a choice of hardware specifications ranging from 2 cores to 32 cores, **although the full range of machine types may not always be available**.

That's super weird, because they are available on another account - what are the rules for product eligibility?

**Comment by sudee404** [comment]
I first used codespaces last year and my quota was over in 2 days. After that, till today it claims I've used 100% of included services despite having no codespaces running for the past 2 months. Check the below screenshots for any ideas about what the problem might be 
![image](https://user-images.githubusercontent.com/58401150/218289767-01a7ff23-93b5-4037-aada-bf73bb245f15.png)

**Reply by hubwriter** [reply]
Did you set up prebuilds? If so, prebuilds will continue to consume storage even if you have no codespaces. 

For more information (including how to check for repos with prebuilds that are consuming your included storage) see this docs article: https://docs.github.com/en/codespaces/troubleshooting/troubleshooting-included-usage#about-codespaces-storage

You can remove prebuilds for repos if you are not actively using Codespaces for those repos and this will avoid using up storage unnecessarily.

**Reply by sudee404** [reply]
No...I don't remember doing anything like that.....I used codespace on one repository for 3 days and I ran out......I then closed and removed it and haven't been able to open another one to this day.....

**Reply by legomushroom** [reply]
@sudee404 seems you have `prebuilds` set up for the `Even******der`(name reduced since this thread is public) repo. The last prebuild was yesterday(Feb 21 2023). Mind checking that on your side?

**Reply by sudee404** [reply]
thanks .... I found it....maybe the reason I couldn't find it was because it was in a private repo

**Reply by subratom** [reply]
@legomushroom where do I find prebuilds? I got a similar 100% usage message and nothing exists under my codespace.

**Comment by m-bruse** [comment]
I am getting the message that I'm out of storage in my codespace.  I am looking to delete what I was doing many months ago and have since forgotten where that is or even what it was called.   Do I need to put my payment information just for the pleasure of deleting that?  or is there some other way?  I'm trying to learn codespaces and probably did not turn it off or delete it when I built it.  

By the way, am I the only one confused by the algorithm for charges?  Seems like an advance degree is needed.  What happened to just displaying what is owed somewhere?  If  I do put my credit card information in how much will I be charged for this experiment I ran?

**Reply by Ishmaello** [reply]
Me I'm just doomed as my financial status is almost non with no earnings, I'm just glad that some solutions regarding codespace is disclosed here. But other than that fees are actually no issue

**Comment by JesusCastroFernandez** [comment]
Codespaces es una herramienta de desarrollo en línea proporcionada por GitHub que permite a los desarrolladores crear, ejecutar y depurar código en un entorno basado en la nube. La versión gratuita de Codespaces tiene algunas limitaciones, pero aún así puedes aprovecharla al máximo. Aquí te dejo algunos consejos:

Aprovecha la capacidad de personalización: Puedes personalizar el entorno de Codespaces para que se adapte a tus necesidades específicas de desarrollo. Por ejemplo, puedes elegir el lenguaje de programación, instalar extensiones, ajustar la configuración de Git y mucho más.

Aprovecha las plantillas predefinidas: Codespaces proporciona plantillas predefinidas para diferentes lenguajes y frameworks de programación, como .NET, Node.js, Python, y Ruby on Rails. Estas plantillas son útiles para empezar rápidamente a desarrollar sin tener que preocuparte por la configuración del entorno.

Utiliza la integración con GitHub: Como Codespaces es una herramienta de GitHub, puedes integrarlo con tu repositorio en GitHub. Esto te permitirá acceder a tu código y trabajar con él directamente desde la interfaz de Codespaces.

Colabora con otros desarrolladores: Codespaces te permite compartir el entorno de desarrollo con otros desarrolladores, lo que es útil si necesitas colaborar en un proyecto. También puedes utilizar Codespaces para realizar sesiones de depuración en tiempo real con otros desarrolladores.

Prueba el rendimiento y la escalabilidad: La versión gratuita de Codespaces tiene algunas limitaciones en cuanto a rendimiento y escalabilidad, pero aún así puedes probar tus aplicaciones y proyectos para ver cómo se comportan en diferentes entornos.

Utiliza los contenedores de Docker: Codespaces utiliza contenedores de Docker para proporcionar el entorno de desarrollo. Esto significa que puedes trabajar en diferentes proyectos en contenedores separados sin tener que preocuparte por la configuración del entorno.

Aprovecha las ventajas de la nube: Al utilizar Codespaces, no necesitas preocuparte por configurar y mantener un entorno de desarrollo local. Todo se realiza en la nube, lo que significa que puedes acceder a tu entorno de desarrollo desde cualquier lugar y en cualquier momento.

En resumen, la versión gratuita de Codespaces es una herramienta muy útil para los desarrolladores que quieren crear, ejecutar y depurar código en un entorno basado en la nube. Con un poco de personalización y aprovechando todas sus funcionalidades, puedes aprovechar al máximo su uso gratuito.

**Comment by noraj** [comment]
I have zero codespace being listed at https://github.com/codespaces.

https://github.com/settings/billing shows I used all the storage and used 0 hour of codespace.

![image](https://github.com/community/community/assets/16578570/9fbf6567-8a0f-45b6-a4cb-70cf8c075c54)

I don't even remember having used codespace this month. What could have consumed all the quota? There is no more details.

Also at https://github.com/settings/codespaces I see I have the Default idle timeout and Default retention period.

**Reply by noraj** [reply]
https://docs.github.com/en/codespaces/troubleshooting/troubleshooting-included-usage#about-codespaces-storage

> The "Your codespaces" page does not list any prebuilds you may have set up. Prebuilds consume storage for a repository, even if you do not currently have any codespaces for that repository. For more information, see "[About GitHub Codespaces prebuilds](https://docs.github.com/en/codespaces/prebuilding-your-codespaces/about-github-codespaces-prebuilds)."

I don't even know what prebuilds are and reading https://docs.github.com/en/codespaces/prebuilding-your-codespaces/about-github-codespaces-prebuilds doesn't help me to find which repository could consume those prebuilds.

**Reply by spenserblack** [reply]
Repositories where you enabled prebuilds (check github.com/\<USERNAME\>/\<REPO\>/settings/codespaces) will start taking up storage. If you disable prebuilds, you should stop taking up prebuild storage. I agree it's hard to find which repos have prebuilds enabled, but you should be able to find which ones are using prebuild storage by generating a usage report from github.com/settings/billing.

One thing to note is that there may be a bug with disabling prebuilds. After disabling them, I was still taking up prebuild storage, and needed someone to manually delete the prebuild: https://github.com/orgs/community/discussions/48213.

**Reply by noraj** [reply]
I found the one but it wasn't easy.

**Comment by RafaelKuhn** [comment]
I also have this problem.

Github keeps telling me "hey you've used N% of the included services for codespaces bla bla bla" (it was growing, until hit 100%)
But I have zero codespaces open, I closed my last one months ago, whenever I enter codespaces it shows nothing, and that crap keeps growing, what the fuck is happening?

I also have never set up any prebuild crap whatsoever.

It really looks like microsoft is has a hand made of ass, EVERYTHING it touches becomes a hunk of liquid shit.

**Reply by rpxgit** [reply]
I second this issue to be addressed please!

**Comment by m-bruse** [comment]
I finally found wher this was hiding.  It was under a repo I had made.

Get the usage report from GITHUB and it will tell you which repo.  From
there you have to delete the pre build.

I would give more detail but honestly it was a hunt and fish method and I
don’t recall where I clicked to delete the prebuild.

**Reply by subratom** [reply]
Thank you. This is the only answer that helped. I looked at my recent repo checkins and clicking on the settings tab -> Codespaces I found that I had a prebuild under specific repo (which should have shown up when I click Codespaces but of course that's not how all of this works, does it) and deleted that pre-build

**Comment by rpxgit** [comment]
![image](https://github.com/community/community/assets/59430183/5e49aaf2-e796-49db-8bd8-010a1d166422)

I keep recieving these and I have no codespaces even opened on my end on my account, I checked my repositories if it is occupying external code storage but found nothing. What is going on?!

**Comment by ahKevinXy** [comment]
I keep recieving these and I have no codespaces even opened on my end on my account, I checked my repositories if it is occupying external code storage but found nothing. What is going on?!

**Comment by hubwriter** [comment]
It's most likely that one or more Codespaces prebuilds are consuming your monthly included storage allowance. You can check this by clicking to expand the "Storage" section of your Codespaces usage details. 

See: 
https://docs.github.com/en/billing/managing-billing-for-github-codespaces/viewing-your-github-codespaces-usage

If this shows use of "Prebuild Storage" you can generate a usage report to find out which repositories have prebuilds set up. See the instructions in the same article. Then go to the settings for those repositories and delete the prebuild configuration. See https://docs.github.com/en/codespaces/prebuilding-your-codespaces/managing-prebuilds#deleting-a-prebuild-configuration.
